lib/schema/array.js

```diff
@@ -393,7 +393,7 @@ SchemaArray.prototype.cast = function(value, doc, init, prev, options) {
     return value;
   }
 
-  if (init || SchemaArray.options.castNonArrays) {
+  if ((init || SchemaArray.options.castNonArrays) && (typeof this.options.castNonArrays === 'boolean' ? this.options.castNonArrays : true)) {
```

Comment:
I think the current state wouldn't work if we have `SchemaArray.options.castNonArrays` set to false, but on a specific path we want to override it to `true`. The following should work.
```suggestion
  const castNonArraysOption = this.options.castNonArrays != null ? this.options.castNonArrays : SchemaArray.options.castNonArrays;
  if (init || castNonArraysOption) {
```

What do you think of writing a test verifying that we can set `SchemaArray.options.castNonArrays` to `false` and then override it on a specific path?

Reply:
currently i have documented that it required both `SchemaArray.options.castNonArrays` set to `true` or `init` AND `castNonArrays` (Property option) unset or set to `true`

Reply:
added a case in b4715b5cb75d3a110e8394b0835798d618327ae9

Reply:
I'm not sure https://github.com/Automattic/mongoose/commit/b4715b5cb75d3a110e8394b0835798d618327ae9 implements what I'm trying to describe, that would look more like that:
```js
it('`castNonArrays` on specific paths takes precedence over global option', function() {
  // Arrange
  const m = new Mongoose();
  m.Schema.Types.Array.options.castNonArrays = false;

  const userSchema = new Schema({ friendsNames: { type: [String], castNonArrays: true } });
  const User = m.model('User', userSchema);

  // Act
  const user = new User({ friendsNames: 'Sam' });

  // Assert
  const err = user.validateSync();
  assert.ok(err == null);
});
```

Reply:
> I'm not sure https://github.com/Automattic/mongoose/commit/b4715b5cb75d3a110e8394b0835798d618327ae9 implements what I'm trying to describe, that would look more like that:

thanks for clarifying, currently b4715b5cb75d3a110e8394b0835798d618327ae9 tests that `SchemaArray.options.castNonArrays` takes priority over the property option

Reply:
I'm a bit lost. Can you please point me to the specific line/column where we set `Schema.Types.Array.options.castNonArrays` to false?
The case that I'd like to cover by a test-case is: globalCastNonArrays: false, propertyCastNonArrays: true.
Does this test-case cover that? 🤨

Reply:
I can see that `beforeEach` and `afterEach` we're re-assigning the values, but I can't see where we _change_ them.

Instead of using beforeEach/afterEach and trying to keep the values unchanged, can we instead instantiate new mongoose instances like https://github.com/Automattic/mongoose/pull/12090#discussion_r917988447? That way the test is self-contained and more explicit.

What do you think?

Reply:
yes this could be done, but before adding i would like to wait for someone like @vkarpov15 to say what direct this change should go:
- let the property option overwrite the global, or
- let the property option only apply when the global is also `true`

Reply:
Going with "path-option overrides global option" is consistent with how we do things in mongoose.
Check the following for example: https://github.com/Automattic/mongoose/blob/db1ba6e2a55a8134b99fc9074597ba2d5219e8b9/lib/helpers/setDefaultsOnInsert.js#L19-L22 
and
https://github.com/Automattic/mongoose/blob/db1ba6e2a55a8134b99fc9074597ba2d5219e8b9/lib/query.js#L3387-L3394

If we go with "letting the property apply _only_ when the global is also true", why would a person specify it if does nothing?

What do you think? @hasezoey @vkarpov15

Reply:
for now i have applied the suggested changes (and the new test), because before i was not sure how mongoose handles such cases, thanks for the examples of where such cases exist already

Reply:
@hasezoey thanks for making the suggested change. The modified behavior that says "use the global value if there's no value set on the current schematype instance" is more consistent with the rest of Mongoose.